### PR TITLE
Andrew/control

### PR DESCRIPTION
**Pull request by exspeed**
Zooming - Use up/down arrow or scroll wheel
Panning - Hold shift then drag (BUG: initial panning doesn't work, calling rotateAround(drag without pressing shift) and then Panning works fine).

### CODE REVIEW

File: src/Camera.cpp [hunk 1]
```diff
@@ -58,3 +58,36 @@ void Camera::RotateAround(float deltaX, float deltaY)
 }
 
 
+void Camera::Zoom(bool in)
+{
+	const float zoomSpeed = 0.2;
+
+	glm::vec3 direction = zoomSpeed*(normalize(forward-pos));
+	if(!in){
```
Comment (ammunoz) [inline src/Camera.cpp]: Left bracket should go to new line to keep consistent with the rest of the code

File: src/Camera.cpp [hunk 2]
```diff
@@ -58,3 +58,36 @@ void Camera::RotateAround(float deltaX, float deltaY)
...
+	}
+	pos = pos + direction;
+
+	UpdateViewMatrix();
+}
+
+void Camera::Panning(float deltaX, float deltaY)
+{
+	const float panningSpeed = 0.1;
+	glm::vec3 bi = normalize(cross(forward-pos, up));
```
Comment (ammunoz) [inline src/Camera.cpp]: These need the namespace glm before the function calls. e.g., glm::normalize(glm::cross(...));

I'm surprised this compiles for you.

File: src/InputManager.cpp [hunk 3]
```diff
@@ -12,10 +12,19 @@ void InputManager::CheckInput()
 	MouseInput();
 }
 
+#include <iostream>
 void InputManager::MouseInput()
 {
 	double xpos, ypos;
-	if(glfwGetMouseButton(window, GLFW_MOUSE_BUTTON_LEFT) == GLFW_PRESS)
+	if( shiftKey && glfwGetMouseButton(window, GLFW_MOUSE_BUTTON_LEFT) == GLFW_PRESS){
```
Comment (ammunoz) [inline src/InputManager.cpp]: Left bracket needs to go to next line to keep consistent with the rest of the code.

File: src/InputManager.cpp [hunk 4]
```diff
@@ -39,4 +48,33 @@ void InputManager::KeyInput()
...
+	}
+	else if(glfwGetKey(window,GLFW_KEY_LEFT_SHIFT) == GLFW_RELEASE)
+	{
+		shiftKey = false;
+	}
+
 }
+
+void InputManager::ScrollWheel(double xoffset, double yoffset){
+    if(yoffset == 1 ){
```
Comment (ammunoz) [inline src/InputManager.cpp]: Left bracket needs to go to next line to keep consistent with the rest of the code.

File: src/InputManager.cpp [hunk 5]
```diff
@@ -39,4 +48,33 @@ void InputManager::KeyInput()
...
+			camera->Zoom(true);
+		}
+    else if (yoffset == -1){
```
Comment (ammunoz) [inline src/InputManager.cpp]: Left bracket needs to go to next line to keep consistent with the rest of the code.

File: src/main.cpp [hunk 6]
```diff
@@ -19,6 +19,17 @@
 #include "ShaderTool.h"
 #include "InputManager.h"
 
+InputManager* inputManager = nullptr;
+
+void scroll_callback(GLFWwindow* window, double xoffset, double yoffset)
+{
+		if(inputManager == NULL){
```
Comment (ammunoz) [inline src/main.cpp]: Left bracket needs to go to next line to keep consistent with the rest of the code.

File: src/main.cpp [hunk 7]
```diff
@@ -19,6 +19,17 @@
...
+			std::cout << "WHOOPS: inputManager is NULL\n";
+			return;
+		}else{
```
Comment (ammunoz) [inline src/main.cpp]: Left bracket needs to go to next line to keep consistent with the rest of the code.

else should go to new line as well.

File: src/Camera.cpp [hunk 8]
```diff
@@ -58,3 +58,36 @@ void Camera::RotateAround(float deltaX, float deltaY)
...
+	if(deltaX > 0)
+		pos = pos - panningSpeed*bi; 		
+	else if(deltaX < 0)
+		pos = pos + panningSpeed*bi; 		
+
+	glm::vec3 updir = normalize(cross(bi, forward-pos));
```
Comment (ammunoz) [inline src/Camera.cpp]: same thing here

File: include/InputManager.h [hunk 9]
```diff
@@ -10,11 +10,13 @@ class InputManager
 	Camera* camera;
 	
 	InputManager(GLFWwindow *window, Camera* cam);
+	void ScrollWheel(double xoffset, double yoffset);
 	void CheckInput();
 	
 private:
 	float cursorX = 0;
 	float cursorY = 0;
+	bool shiftKey = 0;
```
Comment (ammunoz) [inline include/InputManager.h:19]: Why not set this to false as opposed to 0?

**Review by ammunoz — CHANGES_REQUESTED** [review]

File: include/InputManager.h (hunk 9 above)
Comment (exspeed) [reply include/InputManager.h:19]: I don't think it's a big deal. 0/1 or false/true,  this shouldn't be part of our coding standard.

File: src/InputManager.cpp (hunk 3 above)
Comment (exspeed) [reply src/InputManager.cpp]: d

File: src/Camera.cpp (hunk 1 above)
Comment (exspeed) [reply src/Camera.cpp]: d

File: src/Camera.cpp (hunk 2 above)
Comment (exspeed) [reply src/Camera.cpp]: d

File: src/InputManager.cpp (hunk 5 above)
Comment (exspeed) [reply src/InputManager.cpp]: d

File: src/InputManager.cpp (hunk 4 above)
Comment (exspeed) [reply src/InputManager.cpp]: d

File: src/main.cpp (hunk 6 above)
Comment (exspeed) [reply src/main.cpp]: d

File: src/main.cpp (hunk 7 above)
Comment (exspeed) [reply src/main.cpp]: d

File: src/Camera.cpp (hunk 8 above)
Comment (exspeed) [reply src/Camera.cpp]: d

File: include/InputManager.h (hunk 9 above)
Comment (exspeed) [reply include/InputManager.h:19]: d

File: src/InputManager.cpp [hunk 10]
```diff
@@ -39,4 +49,35 @@ void InputManager::KeyInput()
...
+	else if(glfwGetKey(window,GLFW_KEY_LEFT_SHIFT) == GLFW_RELEASE)
+	{
+		shiftKey = false;
+	}
+
 }
+
+void InputManager::ScrollWheel(double xoffset, double yoffset){
+    if(yoffset == 1 )
+		{
```
Comment (DanielDDo) [inline src/InputManager.cpp]: wonky spacing here

**Review by DanielDDo — APPROVED** [review]

File: src/main.cpp [hunk 11]
```diff
@@ -19,6 +19,20 @@
 #include "ShaderTool.h"
 #include "InputManager.h"
 
+InputManager* inputManager = nullptr;
+
+void scroll_callback(GLFWwindow* window, double xoffset, double yoffset)
+{
+		if(inputManager == NULL)
```
Comment (ammunoz) [inline src/main.cpp]: NULL is deprecated. Use nullptr

Fix this and merge in

**Review by ammunoz — APPROVED** [review]

**Pull request merged**